File: DLT+P3P/src/core/block.py
import xml.etree.ElementTree as ET
import numpy as np
from typing import List, Tuple, Optional
from ..structures.camera import Camera
from ..structures.matrices import Mat_K, Mat_R
from ..structures.vectors import Vec_t
from ..structures.features import TiePnt, Color, Obser
from ..structures.photo import Photo
from ..utils.transformations import rotation_matrix_to_euler_angles
from ..utils.pose_estimation import (
    dlt_pose_estimation, 
    p3p_pose_estimation, 
    ransac_pose_estimation,
    undistort_points
)

class Block:
    def __init__(self):
        self.camera = Camera(Mat_K())
        self.tiepoints: List[TiePnt] = []
        
    def load_from_xml(self, filename: str) -> bool:
        try:
            tree = ET.parse(filename)
            root = tree.getroot()
            block_elem = root.find('Block')
            if block_elem is None:
                return False
                
            # 读取相机参数
            photogroups = block_elem.find('Photogroups')
            if photogroups is not None:
                photogroup = photogroups.find('Photogroup')
                if photogroup is not None:
                    # 读取内参
                    f = float(photogroup.find('FocalLength').text)
                    pp = photogroup.find('PrincipalPoint')
                    cx = float(pp.find('x').text)
                    cy = float(pp.find('y').text)
                    
                    # 设置相机内参矩阵
                    self.camera.intrinsic = Mat_K(np.array([
                        [f, 0, cx],
                        [0, f, cy],
                        [0, 0, 1]
                    ]))
                    
                    # 读取畸变参数
                    distortion = photogroup.find('Distortion')
                    if distortion is not None:
                        self.camera.k1 = float(distortion.find('K1').text)
                        self.camera.k2 = float(distortion.find('K2').text)
                        self.camera.p1 = float(distortion.find('P1').text)
                        self.camera.p2 = float(distortion.find('P2').text)
                    
                    # 读取照片信息
                    for photo in photogroup.findall('Photo'):
                        photo_data = self._parse_photo(photo)
                        self.camera.photos.append(photo_data)
            
            # 读取特征点
            tiepoints_elem = block_elem.find('TiePoints')
            if tiepoints_elem is not None:
                for tiepoint in tiepoints_elem.findall('TiePoint'):
                    tp = self._parse_tiepoint(tiepoint)
                    self.tiepoints.append(tp)
                    
            return True
            
        except Exception as e:
            print(f"加载XML文件失败: {str(e)}")
            return False
    
    def _parse_photo(self, photo_elem) -> Photo:
        # 解析照片信息的辅助方法
        image_path = photo_elem.find('ImagePath').text
        label = image_path.split('/')[-1]
        
        pose = photo_elem.find('Pose')
        center = pose.find('Center')
        position = Vec_t(
            float(center.find('x').text),
            float(center.find('y').text),
            float(center.find('z').text)
        )
        
        rotation = pose.find('Rotation')
        rot_matrix = np.zeros((3, 3))
        for i in range(3):
            for j in range(3):
                elem = rotation.find(f'M_{i}{j}')
                rot_matrix[i][j] = float(elem.text)
                
        # 使用工具函数计算欧拉角
        euler_angles = rotation_matrix_to_euler_angles(rot_matrix)
        
        return Photo(
            label=label,
            position=position,
            rotation=Mat_R(rot_matrix),
            euler_angles=euler_angles
        ) 
    
    def _parse_tiepoint(self, tiepoint_elem) -> TiePnt:
        """解析特征点信息"""
        position = tiepoint_elem.find('Position')
        pos = Vec_t(
            float(position.find('x').text),
            float(position.find('y').text),
            float(position.find('z').text)
        )
        
        color_elem = tiepoint_elem.find('Color')
        color = None
        if color_elem is not None:
            # 将0-1范围的浮点数转换为0-255范围的整数
            r = int(float(color_elem.find('Red').text) * 255)
            g = int(float(color_elem.find('Green').text) * 255)
            b = int(float(color_elem.find('Blue').text) * 255)
            color = Color(r, g, b)
            
        tp = TiePnt(position=pos, color=color)
        
        # 解析观测点
        for measurement in tiepoint_elem.findall('Measurement'):
            photo_id = int(measurement.find('PhotoId').text)
            x = float(measurement.find('x').text)
            y = float(measurement.find('y').text)
            tp.observations.append(Obser(photo_id, x, y))
            
        return tp

    def get_corresponding_points(self, photo_id: int) -> Tuple[List[Vec_t], List[Obser]]:
        """获取指定照片的3D点和对应的2D观测点"""
        points_3d = []
        points_2d = []
        
        for tp in self.tiepoints:
            for obs in tp.observations:
                if obs.photo_id == photo_id:
                    points_3d.append(tp.position)
                    points_2d.append(obs)
                    break
                    
        return points_3d, points_2d

    def estimate_pose_dlt(self, photo_id: int) -> Optional[Tuple[Mat_R, Vec_t, float]]:
        """使用DLT估计相机位姿"""
        points_3d, points_2d = self.get_corresponding_points(photo_id)
        if len(points_3d) < 6:
            print(f"点数不足，需要至少6个点，当前只有{len(points_3d)}个点")
            return None
            
        # 获取相机参数
        K = self.camera.intrinsic.data
        
        # 使用原始点进行DLT估计
        print(f"使用{len(points_3d)}个点进行DLT估计")
        return dlt_pose_estimation(points_3d, points_2d, K)

    def estimate_pose_p3p(self, photo_id: int) -> Tuple[Optional[Mat_R], Optional[Vec_t]]:
        """使用P3P估计相机位姿"""
        points_3d, points_2d = self.get_corresponding_points(photo_id)
        if len(points_3d) < 4:  # P3P需要至少4个点
            print(f"点数不足，需要至少4个点，当前只有{len(points_3d)}个点")
            return None, None

        # 取相机参数
        K = self.camera.intrinsic.data
        
        return p3p_pose_estimation(points_3d, points_2d, K)

    def estimate_pose_ransac(self, photo_id: int, method: str = 'dlt') -> Optional[Tuple[Mat_R, Vec_t, List[bool], float]]:
        """使用RANSAC估计相机位姿"""
        points_3d, points_2d = self.get_corresponding_points(photo_id)
        
        if len(points_3d) < 6:
            print(f"点数不足，需要至少6个点，当前只有{len(points_3d)}个点")
            return None
        
        # 获取相机参数
        K = self.camera.intrinsic.data
        
        # 调用RANSAC位姿估计
        print(f"使用{len(points_3d)}个点进行RANSAC {method}估计")
        return ransac_pose_estimation(
            points_3d=points_3d,
            points_2d=points_2d,
            K=K,
            method=method,
            ransac_iterations=200,  # 增加迭代次数
            ransac_threshold=500.0  # 设置更大的阈值
        )

File: src/types/block.cpp
#include "../../include/types/block.hpp"
#include <iostream>
#include <tinyxml2.h>
#include <cmath>  // 添加这个用于atan2和sqrt函数
namespace slam {
    // 在Block类的实现之前添加函数定义
    Vec_t rotationMatrixToEulerAngles(
        double r11, double r12, double r13,
        double r21, double r22, double r23,
        double r31, double r32, double r33) {

        Vec_t angles;
        angles.x = atan2(r32, r33);
        angles.y = atan2(-r31, sqrt(r32 * r32 + r33 * r33));
        angles.z = atan2(r21, r11);
        return angles;
    }


// Block类XML读写实现
bool Block::loadFromXML(const std::string& filename) {
    tinyxml2::XMLDocument doc;
    if (doc.LoadFile(filename.c_str()) != tinyxml2::XML_SUCCESS) {
        std::cout << "XML文件加载失败: " << filename << std::endl;
        std::cout << "错误信息: " << doc.ErrorStr() << std::endl;
        return false;
    }

    tinyxml2::XMLElement* root = doc.RootElement();
    if (!root) return false;

    tinyxml2::XMLElement* blockElem = root->FirstChildElement("Block");
    if (!blockElem) return false;

    // 读取相机参数
    tinyxml2::XMLElement* photogroups = blockElem->FirstChildElement("Photogroups");
    if (photogroups) {
        tinyxml2::XMLElement* photogroup = photogroups->FirstChildElement("Photogroup");
        if (photogroup) {
            // 读取图像尺寸
            tinyxml2::XMLElement* imageDimensions = photogroup->FirstChildElement("ImageDimensions");
            if (imageDimensions) {
                tinyxml2::XMLElement* widthElem = imageDimensions->FirstChildElement("Width");
                tinyxml2::XMLElement* heightElem = imageDimensions->FirstChildElement("Height");
                // 可以存储尺寸信息如果需要的话
            }

            // 读取内参
            tinyxml2::XMLElement* focalLength = photogroup->FirstChildElement("FocalLength");
            double f = focalLength ? focalLength->DoubleText() : 0.0;

            // 读取主点
            tinyxml2::XMLElement* principalPoint = photogroup->FirstChildElement("PrincipalPoint");
            double cx = 0.0, cy = 0.0;
            if (principalPoint) {
                tinyxml2::XMLElement* x = principalPoint->FirstChildElement("x");
                tinyxml2::XMLElement* y = principalPoint->FirstChildElement("y");
                if (x) cx = x->DoubleText();
                if (y) cy = y->DoubleText();
            }

            // 设置相机内参矩阵
            camera.intrinsic.data[0][0] = f;  // fx
            camera.intrinsic.data[0][1] = 0;
            camera.intrinsic.data[0][2] = cx;
            camera.intrinsic.data[1][0] = 0;
            camera.intrinsic.data[1][1] = f;  // fy
            camera.intrinsic.data[1][2] = cy;
            camera.intrinsic.data[2][0] = 0;
            camera.intrinsic.data[2][1] = 0;
            camera.intrinsic.data[2][2] = 1;

            // 读取畸变系数
            tinyxml2::XMLElement* distortion = photogroup->FirstChildElement("Distortion");
            if (distortion) {
                tinyxml2::XMLElement* k1Elem = distortion->FirstChildElement("K1");
                tinyxml2::XMLElement* k2Elem = distortion->FirstChildElement("K2");
                tinyxml2::XMLElement* p1Elem = distortion->FirstChildElement("P1");
                tinyxml2::XMLElement* p2Elem = distortion->FirstChildElement("P2");

                if (k1Elem) camera.k1 = k1Elem->DoubleText();
                if (k2Elem) camera.k2 = k2Elem->DoubleText();
                if (p1Elem) camera.p1 = p1Elem->DoubleText();
                if (p2Elem) camera.p2 = p2Elem->DoubleText();
            }

            // 读取照片数据
            for (tinyxml2::XMLElement* photo = photogroup->FirstChildElement("Photo");
                photo;
                photo = photo->NextSiblingElement("Photo")) {

                Photo photoData;

                // 读取图像路径
                tinyxml2::XMLElement* imagePath = photo->FirstChildElement("ImagePath");
                if (imagePath && imagePath->GetText()) {
                    std::string path = imagePath->GetText();
                    size_t lastSlash = path.find_last_of("/\\");
                    photoData.label = (lastSlash != std::string::npos) ? path.substr(lastSlash + 1) : path;
                }

                // 读取位姿
                tinyxml2::XMLElement* poseElem = photo->FirstChildElement("Pose");
                if (poseElem) {
                    // 读取中心点位置
                    tinyxml2::XMLElement* center = poseElem->FirstChildElement("Center");
                    if (center) {
                        tinyxml2::XMLElement* x = center->FirstChildElement("x");
                        tinyxml2::XMLElement* y = center->FirstChildElement("y");
                        tinyxml2::XMLElement* z = center->FirstChildElement("z");

                        if (x && y && z) {
                            photoData.position = Vec_t(
                                x->DoubleText(),
                                y->DoubleText(),
                                z->DoubleText()
                            );
                        }
                    }

                    // 读取旋转矩阵
                    tinyxml2::XMLElement* rotation = poseElem->FirstChildElement("Rotation");
                    if (rotation) {
                        const char* elements[] = {
                            "M_00", "M_01", "M_02",
                            "M_10", "M_11", "M_12",
                            "M_20", "M_21", "M_22"
                        };

                        for (int i = 0; i < 3; i++) {
                            for (int j = 0; j < 3; j++) {
                                tinyxml2::XMLElement* elem = rotation->FirstChildElement(elements[i * 3 + j]);
                                if (elem) {
                                    photoData.rotation.data[i][j] = elem->DoubleText();
                                }
                            }
                        }

                        // 计算欧拉角
                        photoData.euler_angles = rotationMatrixToEulerAngles(
                            photoData.rotation.data[0][0], photoData.rotation.data[0][1], photoData.rotation.data[0][2],
                            photoData.rotation.data[1][0], photoData.rotation.data[1][1], photoData.rotation.data[1][2],
                            photoData.rotation.data[2][0], photoData.rotation.data[2][1], photoData.rotation.data[2][2]
                        );
                    }
                }

                photoData.enabled = true;
                camera.photos.push_back(photoData);
            }
        }
    }

    // 读取特征点
    tinyxml2::XMLElement* tiePointsElem = blockElem->FirstChildElement("TiePoints");
    if (tiePointsElem) {
        for (tinyxml2::XMLElement* tiePoint = tiePointsElem->FirstChildElement("TiePoint");
            tiePoint;
            tiePoint = tiePoint->NextSiblingElement("TiePoint")) {

            TiePnt tp;

            // 读取位置
            tinyxml2::XMLElement* position = tiePoint->FirstChildElement("Position");
            if (position) {
                tinyxml2::XMLElement* x = position->FirstChildElement("x");
                tinyxml2::XMLElement* y = position->FirstChildElement("y");
                tinyxml2::XMLElement* z = position->FirstChildElement("z");

                if (x && y && z) {
                    tp.position = Vec_t(
                        x->DoubleText(),
                        y->DoubleText(),
                        z->DoubleText()
                    );
                }
            }

            // 读取颜色
            tinyxml2::XMLElement* color = tiePoint->FirstChildElement("Color");
            if (color) {
                tinyxml2::XMLElement* r = color->FirstChildElement("Red");
                tinyxml2::XMLElement* g = color->FirstChildElement("Green");
                tinyxml2::XMLElement* b = color->FirstChildElement("Blue");

                if (r && g && b) {
                    tp.color = Color(
                        r->DoubleText(),
                        g->DoubleText(),
                        b->DoubleText()
                    );
                }
            }

            // 读取观测点
            for (tinyxml2::XMLElement* measurement = tiePoint->FirstChildElement("Measurement");
                measurement;
                measurement = measurement->NextSiblingElement("Measurement")) {

                tinyxml2::XMLElement* photoId = measurement->FirstChildElement("PhotoId");
                tinyxml2::XMLElement* x = measurement->FirstChildElement("x");
                tinyxml2::XMLElement* y = measurement->FirstChildElement("y");

                if (photoId && x && y) {
                    tp.observations.push_back(Obser(
                        photoId->IntText(),
                        x->DoubleText(),
                        y->DoubleText()
                    ));
                }
            }

            tiepoints.push_back(tp);
        }
    }

    return true;
}

bool Block::saveToXML(const std::string& filename) const {
    tinyxml2::XMLDocument doc;

    // 创建根元素
    tinyxml2::XMLElement* root = doc.NewElement("ContextCapture");
    doc.InsertFirstChild(root);

    // 创建Block元素
    tinyxml2::XMLElement* blockElem = doc.NewElement("Block");
    root->InsertEndChild(blockElem);

    // 保存相机参数
    tinyxml2::XMLElement* photogroups = doc.NewElement("Photogroups");
    blockElem->InsertEndChild(photogroups);

    tinyxml2::XMLElement* photogroup = doc.NewElement("Photogroup");
    photogroups->InsertEndChild(photogroup);

    // 保存内参
    tinyxml2::XMLElement* focalLength = doc.NewElement("FocalLength");
    focalLength->SetText(camera.intrinsic.data[0][0]);
    photogroup->InsertEndChild(focalLength);

    // ... 保存其他参数 ...

    return doc.SaveFile(filename.c_str()) == tinyxml2::XML_SUCCESS;
}

} // namespace slam 